apollo-router/src/plugins/telemetry/config_new/cost/mod.rs

```diff
@@ -49,14 +49,23 @@ pub(crate) struct SupergraphCostAttributes {
 impl Selectors for SupergraphCostAttributes {
     type Request = supergraph::Request;
     type Response = supergraph::Response;
+    type EventResponse = crate::graphql::Response;
 
     fn on_request(&self, _request: &Self::Request) -> Vec<KeyValue> {
         Vec::default()
     }
 
-    fn on_response(&self, response: &Self::Response) -> Vec<KeyValue> {
+    fn on_response(&self, _response: &Self::Response) -> Vec<KeyValue> {
+        Vec::default()
+    }
+
+    fn on_error(&self, _error: &BoxError) -> Vec<KeyValue> {
+        Vec::default()
+    }
+
+    fn on_event_response(&self, _response: &Self::EventResponse, ctx: &Context) -> Vec<KeyValue> {
         let mut attrs = Vec::with_capacity(4);
-        if let Some(cost_result) = &response.context.extensions().lock().get::<CostContext>() {
+        if let Some(cost_result) = &ctx.extensions().lock().get::<CostContext>() {
```

Comment:
I think we're holding the lock for longer than we should.

Reply:
I don't think this is the case.

There's nothing blocking or async in the block below, so we could obtain the lock for each attribute, but the overhead of repeatedly locking could be worse.

Reply:
we aren't locking repeatedly, we're cloning CostContext and releasing the lock immediately instead.